### PR TITLE
fs: add uv_fs_mkstemp

### PR DESCRIPTION
**Pull request by saghul**
Supersedes: https://github.com/libuv/libuv/pull/2074

### CODE REVIEW

**Comment by saghul** [comment]
TODO:

- Windows (I'll try to get that done tonight)

Changes from. original PR:

- address the feedback I provided back in the day
- remove O_LARGEFILE and mkostemp64 handling, it should be automatic AFAICT since we build with `_LARGEFILE_SOURCE` and `_FILE_OFFSET_BITS=64` defined.

File: docs/src/fs.rst [hunk 1]
```diff
@@ -245,10 +246,14 @@ API
...
+    Equivalent to :man:`mkdtemp(3)`. The result can be found as a null terminated string at `req->path`.
 
-    .. note::
-        The result can be found as a null terminated string at `req->path`.
+.. c:function:: int uv_fs_mkstemp(uv_loop_t* loop, uv_fs_t* req, const char* tpl, uv_fs_cb cb)
+
+    Equivalent to :man:`mkstemp(3)`. The created file path can be found as a null terminated string at `req->path`.
+    The file descriptor can be found as an integer at `req->result`.
+
+    .. versionadded:: 1.35.0
```
Comment (cjihrig) [inline docs/src/fs.rst]: This could still make it into 1.34.0.

Comment (saghul) [reply docs/src/fs.rst]: Sweet. I'll try to get it all ready tonight.

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1647/

**Comment by saghul** [comment]
This one is now ready for review.

**Comment by musm** [comment]
How is a unique temporary filename generated on Windows in this PR?

**Comment by saghul** [comment]
@musm It reuses the same logic as mkdtemp, which was implemented before. It uses the system random generator. The implementation was adapted from OpenBSD. See fs__mktemp.

File: src/unix/fs.c [hunk 2]
```diff
@@ -258,6 +259,80 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
...
 
 
+static int uv__fs_mkstemp(uv_fs_t* req) {
+  int r;
+#ifdef O_CLOEXEC
+  int (*mkostemp_function)(char*, int);
+  static int no_cloexec_support;
+#endif
+  static const char pattern[] = "XXXXXX";
+  static const size_t pattern_size = 6;
```
Comment (bnoordhuis) [inline src/unix/fs.c]:
```suggestion
  static const size_t pattern_size = sizeof(pattern) - 1;
```

File: src/unix/fs.c [hunk 3]
```diff
@@ -258,6 +259,80 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
...
+
+    if (mkostemp_function != NULL) {
+      r = mkostemp_function(path, O_CLOEXEC);
+
+      if (r >= 0)
+        return r;
+
+      /* If mkostemp() returns EINVAL, it means the kernel doesn't
+         support O_CLOEXEC, so we just fallback to mkstemp() below. */
+      if (r == -1 && errno != EINVAL)
```
Comment (bnoordhuis) [inline src/unix/fs.c]:
```suggestion
      if (errno != EINVAL)
```
(the `r == -1` check is superfluous, right?)

File: src/unix/fs.c [hunk 4]
```diff
@@ -258,6 +259,80 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
...
+        return r;
+
+      /* We set the static variable so that next calls don't even
+         try to use mkostemp. */
+      no_cloexec_support = 1;
+    }
+  }
+#endif
```
Comment (bnoordhuis) [inline src/unix/fs.c]:
```suggestion
#endif  /* O_CLOEXEC */
```

File: test/test-fs.c [hunk 5]
```diff
@@ -1208,6 +1238,69 @@ TEST_IMPL(fs_mkdtemp) {
 }
 
 
+TEST_IMPL(fs_mkstemp) {
+  int r;
+  int fd;
+  const char* path_template = "test_file_XXXXXX";
```
Comment (bnoordhuis) [inline test/test-fs.c]:
```suggestion
  const char path_template[] = "test_file_XXXXXX";
```

File: src/unix/fs.c (hunk 3 above)
Comment (saghul) [reply src/unix/fs.c]: right

File: src/unix/fs.c (hunk 2 above)
Comment (saghul) [reply src/unix/fs.c]: 👍

File: test/test-fs.c (hunk 5 above)
Comment (saghul) [reply test/test-fs.c]: 👍

**Comment by saghul** [comment]
Thanks for the suggestions @bnoordhuis added them!

File: src/unix/fs.c [hunk 6]
```diff
@@ -258,6 +259,80 @@ static ssize_t uv__fs_mkdtemp(uv_fs_t* req) {
...
 
 
+static int uv__fs_mkstemp(uv_fs_t* req) {
+  int r;
+#ifdef O_CLOEXEC
+  int (*mkostemp_function)(char*, int);
+  static int no_cloexec_support;
+#endif
+  static const char pattern[] = "XXXXXX";
+  static const size_t pattern_size = sizeof(pattern) -1;
```
Comment (cjihrig) [inline src/unix/fs.c]: Super tiny nit:

```suggestion
  static const size_t pattern_size = sizeof(pattern) - 1;
```

File: src/unix/fs.c [hunk 7]
```diff
@@ -1866,10 +1954,12 @@ void uv_fs_req_cleanup(uv_fs_t* req) {
 
   /* Only necessary for asychronous requests, i.e., requests with a callback.
    * Synchronous ones don't copy their arguments and have req->path and
-   * req->new_path pointing to user-owned memory.  UV_FS_MKDTEMP is the
-   * exception to the rule, it always allocates memory.
+   * req->new_path pointing to user-owned memory.  UV_FS_MKDTEMP and 
+   * UV_FS_MKSTEMP are the exception to the rule, they always allocates memory.
```
Comment (cjihrig) [inline src/unix/fs.c]:
```suggestion
   * UV_FS_MKSTEMP are the exception to the rule, they always allocate memory.
```

File: src/win/fs.c [hunk 8]
```diff
@@ -1244,6 +1243,78 @@ void fs__mkdtemp(uv_fs_t* req) {
 }
 
 
+static int fs__mkdtemp_func(uv_fs_t* req) {
+  if (_wmkdir(req->file.pathw) == 0) {
+    SET_REQ_RESULT(req, 0);
+    return 1;
+  } else if (errno != EEXIST) {
+    SET_REQ_RESULT(req, -1);
+    SET_REQ_RESULT(req, -1);
```
Comment (cjihrig) [inline src/win/fs.c]: Duplicate lines?

File: src/win/fs.c [hunk 9]
```diff
@@ -1244,6 +1243,78 @@ void fs__mkdtemp(uv_fs_t* req) {
...
+  }
+
+  SET_REQ_RESULT(req, fd);
+
+  return 1;
+}
+
+
+void fs__mkstemp(uv_fs_t* req) {
+  fs__mktemp(req,  fs__mkstemp_func);
```
Comment (cjihrig) [inline src/win/fs.c]:
```suggestion
  fs__mktemp(req, fs__mkstemp_func);
```

**Comment by saghul** [comment]
Thanks for the review @cjihrig!  Fixed all the nits and landed!

**Pull request merged**